File: src/Messages/GameMessages.cs
﻿using System;
using System.Collections.Generic;
using System.Diagnostics;
using Akka.Actor;
using Messages.CSharp.Containers;
using Messages.CSharp.Pieces;

namespace Messages.CSharp
{
    [DebuggerStepThrough]
    public abstract partial class Message
    {

        public class CreateGame : WithToken
        {
            public CreateGame(Guid token)
                : base(token)
            {
            }
        }

        public class UnableToCreateGame : WithToken
        {
            public string Error { get; private set; }

            public UnableToCreateGame(Guid token, string error) : base(token)
            {
                Error = error;
            }
        }

        public class PlayerArrived
        {
            public ActorInfoContainer Player { get; private set; }

            public PlayerArrived(ActorInfoContainer player)
            {
                Player = player;
            }
        }

        public class GameStatusUpdate : GameMessage
        {
            public IActorRef Game { get; private set; }

            public GameStatus Status { get; private set; }

            public string Message { get; private set; }

            public GameStatusUpdate(Guid token, Guid gameToken, GameStatus status, IActorRef game,
                string message = null) : base(token, gameToken)
            {
                Status = status;
                Game = game;
                Message = message;
            }
        }

        public class PlayerJoining : GameMessage
        {
            public ActorInfoContainer Player { get; private set; }

            public PlayerJoining(Guid gameToken, ActorInfoContainer player)
                : base(player.Token, gameToken)
            {
                Player = player;
            }
        }

        public abstract class GameMessage : WithToken
        {
            public Guid GameToken { get; private set; }

            protected GameMessage(Guid token, Guid gameToken) : base(token)
            {
                GameToken = gameToken;
            }
        }

        public class ShipPosition : GameMessage
        {
            public Ship Ship { get; private set; }

            public ShipPosition(Guid token, Guid gameToken, Ship ship)
                : base(token, gameToken)
            {
                Ship = ship;
            }
        }

        public class GiveMeNextPosition : GameMessage
        {
            public Tuple<string, int> Config { get; private set; }

            public string ErrorInPreviousConfig { get; private set; }

            public GiveMeNextPosition(Guid token, Guid gameToken, Tuple<string, int> config, string error)
                : base(token, gameToken)
            {
                Config = config;
                ErrorInPreviousConfig = error;
            }
        }

        public class Missile : GameMessage
        {
            public Point Point { get; private set; }

            public Missile(Guid token, Guid gameToken, Point point)
                : base(token, gameToken)
            {
                Point = point;
            }
        }

        public class GameTable : GameMessage
        {
            public IReadOnlyList<Point> Points { get; private set; }

            public GameTable(Guid token, Guid gameToken, IReadOnlyList<Point> points)
                : base(token, gameToken)
            {
                Points = points;
            }
        }

        public class MissileAlreadyHit : GameMessage
        {
            public Point Point { get; private set; }

            public MissileAlreadyHit(Guid token, Guid gameToken, Point point)
                : base(token, gameToken)
            {
                Point = point;
            }
        }

        public class MissileWasAHit : WithPoint
        {
            public bool ShipDestroyed { get; private set; }

            public MissileWasAHit(Guid token, Guid gameToken, Point point, bool shipDestroyed = false)
                : base(token, gameToken, point)
            {
                ShipDestroyed = shipDestroyed;
            }

        }

        public class StopGame : GameMessage
        {
            public StopGame(Guid userToken, Guid gameToken) : base(userToken, gameToken)
            {
            }
        }
    }
}

File: src/Messages.FSharp/GameMessages.fs
namespace Messages.FSharp
    open System
    open Akka.Actor
    open Messages.FSharp
    open Messages.FSharp.TablesAndShips
    open System.Collections.Generic

    [<AbstractClass>]
    type GameMessage(token, gameToken: Guid) =
        inherit WithToken(token)
        member this.GameToken = gameToken
    
    [<AbstractClass>]
    type WithPoint(token, gameToken, point:Point) = 
        inherit GameMessage(token, gameToken)
        member this.Point = point

    type CreateGame(token: System.Guid) =
        inherit WithToken(token)

    type UnableToCreateGame(token, error:string) =
        inherit WithToken(token)
        member this.Error = error
    
    type PlayerArrived(player:ActorInfoContainer) =
        member this.Player = player

    type GameStatusUpdate(token, gameToken, status:GameStatus, game:IActorRef, message:string) =
        inherit GameMessage(token, gameToken)
        member this.Game = game
        member this.Status = status
        member this.Message = message

    type PlayerJoining(token, gameToken, player:ActorInfoContainer) =
        inherit GameMessage(token, gameToken)
        member this.Player = player

    type ShipPosition(token, gameToken, ship:Ship) =
        inherit GameMessage(token, gameToken)
        member this.Ship = ship

    type GiveMeNextPosition(token, gameToken, config:Piece, error:string) =
        inherit GameMessage(token , gameToken)
        member this.Config = config
        member this.ErrorInPreviousConfig = error

    type Missile(token, gameToken, point) =
        inherit WithPoint(token, gameToken, point)

    type GameTable(token, gameToken, points:IReadOnlyList<Point>) = 
        inherit GameMessage(token, gameToken)
        member this.Points = points

    type MissileAlreadyHit(token, gameToken, point) = 
        inherit WithPoint(token, gameToken, point)

    type MissileWasAHit(token, gameToken, point, shipDestroyed:bool) = 
        inherit WithPoint(token, gameToken, point)
        member this.ShipDestroyed = shipDestroyed

    type StopGame(userToken, gameToken) = 
        inherit GameMessage(userToken, gameToken)
